File: A06 Nodes Related/002 Binary Tree/LC257 Binary Tree Path/LC257 BinaryTreePath.java
##########Fast solution#####
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    List<String> ans;
    public List<String> binaryTreePaths(TreeNode root) {
        ans = new ArrayList<>();
        solve(root, new StringBuilder());
        return ans;
    }
    
    public void solve(TreeNode root, StringBuilder res){
        if(root == null)
            return;
        
        int len = res.length();
        res.append(root.val);
        if(root.left == null && root.right == null){
            ans.add(res.toString());
        }
        else{
            res.append("->");
            if(root.left != null)
                solve(root.left, res);
            if(root.right != null)
                solve(root.right, res);
        }
        res.setLength(len);
    }
}


####### Recursively#######
public class Solution {

##########################
//Recursion
public List<String> binaryTreePaths(TreeNode root) {
    List<String> sList=new LinkedList<String>();
    //String s=new String();
    if (root==null) return sList;
    if (root.left==null && root.right==null) {
        sList.add(Integer.toString(root.val));
        return sList;
    }
    
    for (String s: binaryTreePaths(root.left)) {
        sList.add(Integer.toString(root.val)+"->"+s);
    }
    for (String s: binaryTreePaths(root.right)) {
        sList.add(Integer.toString(root.val)+"->"+s);
    }
    return sList;
}
#############################


# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def binaryTreePaths(self, root: Optional[TreeNode]) -> List[str]:
        if not root:
            return []
        res = []
        self.dfs (root, "", res)
        return res 
    
    def dfs(self, root, ls, res):
        if not root.left and not root.right: 
            res.append(ls + str(root.val))
        if root.left: 
            self.dfs(root.left, ls+ str(root.val) +"->", res)
        if root.right:
            self.dfs(root.right,ls+ str(root.val) +"->", res)



############## dfs + stack####################
public class Solution {
//DFS - Stack
public List<String> binaryTreePaths(TreeNode root) {
    List<String> list=new ArrayList<String>();
    Stack<TreeNode> sNode=new Stack<TreeNode>();
    Stack<String> sStr=new Stack<String>();
    
    if(root==null) return list;
    sNode.push(root);
    sStr.push("");
    while(!sNode.isEmpty()) {
        TreeNode curNode=sNode.pop();
        String curStr=sStr.pop();
        
        if(curNode.left==null && curNode.right==null) list.add(curStr+curNode.val);
        if(curNode.left!=null) {
            sNode.push(curNode.left);
            sStr.push(curStr+curNode.val+"->");
        }
        if(curNode.right!=null) {
            sNode.push(curNode.right);
            sStr.push(curStr+curNode.val+"->");
        }
    }
    return list;
}

############### bfs + queue############################

    public class Solution {
//BFS - Queue
public List<String> binaryTreePaths(TreeNode root) {
    List<String> list=new ArrayList<String>();
    Queue<TreeNode> qNode=new LinkedList<TreeNode>();
    Queue<String> qStr=new LinkedList<String>();
    
    if (root==null) return list;
    qNode.add(root);
    qStr.add("");
    while(!qNode.isEmpty()) {
        TreeNode curNode=qNode.remove();
        String curStr=qStr.remove();
        
        if (curNode.left==null && curNode.right==null) list.add(curStr+curNode.val);
        if (curNode.left!=null) {
            qNode.add(curNode.left);
            qStr.add(curStr+curNode.val+"->");
        }
        if (curNode.right!=null) {
            qNode.add(curNode.right);
            qStr.add(curStr+curNode.val+"->");
        }
    }
    return list;
}

File: A06 Nodes Related/002 Binary Tree/LC257 Binary Tree Path/LC257 BinaryTreePath.py
#######one fast solution#####
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    # Recursive
    # Linear time and space
    def binaryTreePaths(self, root):
        if not root:
            return []
        
        # Check leaf
        if not root.left and not root.right:
            return [str(root.val)]
        
        leftPath = self.binaryTreePaths(root.left)
        rightPath = self.binaryTreePaths(root.right)
        
        leftPath = [str(root.val) + '->' + path for path in leftPath]
        rightPath = [str(root.val) + '->' + path for path in rightPath]
        
        return leftPath + rightPath
    



#######DFS Recursively#######
Runtime: 64 ms, faster than 5.09% of Python3 online submissions for Binary Tree Paths.
#############################


# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def binaryTreePaths(self, root: Optional[TreeNode]) -> List[str]:
        if not root:
            return []
        res = []
        self.dfs (root, "", res)
        return res 
    
    def dfs(self, root, ls, res):
        if not root.left and not root.right: 
            res.append(ls + str(root.val))
        if root.left: 
            self.dfs(root.left, ls+ str(root.val) +"->", res)
        if root.right:
            self.dfs(root.right,ls+ str(root.val) +"->", res)



############## dfs + stack####################
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def binaryTreePaths(self, root: Optional[TreeNode]) -> List[str]:
        if not root: 
            return []
        res, stack = [], [(root, "")]
        
        while stack: 
            node, ls = stack.pop()
            if not node.left and not node.right:
                res.append(ls + str(node.val))
            if node.right: 
                stack.append((node.right, ls+str(node.val) + "->"))
            if node.left: 
                stack.append((node.left, ls +str(node.val) + "->"))
        return res

############### bfs + queue############################

    def binaryTreePaths2(self, root):
        if not root:
            return []
        res, queue = [], collections.deque([(root, "")])
        while queue:
            node, ls = queue.popleft()
            if not node.left and not node.right:
                res.append(ls+str(node.val))
            if node.left:
                queue.append((node.left, ls+str(node.val)+"->"))
            if node.right:
                queue.append((node.right, ls+str(node.val)+"->"))
        return res